src/LinuxTracing/LeafFunctionCallManager.cpp

```diff
@@ -69,37 +75,46 @@ Callstack::CallstackType LeafFunctionCallManager::PatchCallerOfLeafFunction(
     return Callstack::kComplete;
   }
 
-  const uint64_t stack_size = rbp - rsp;
-
+  // Perform one unwinding step. We will only need the memory from $rbp to $rsp + 16 (ensure to
+  // include the previous frame pointer and the return address) for unwinding. If $rbp does not
+  // change from unwinding, we need to patch in the pc after unwinding.
+  const uint64_t stack_size = rbp - rsp + 16;
   const LibunwindstackResult& libunwindstack_result = unwinder->Unwind(
       event_data->pid, current_maps->Get(), event_data->GetRegisters(), event_data->GetStackData(),
-      std::min<uint64_t>(stack_size, stack_dump_size_), true);
-  const std::vector<unwindstack::FrameData>& libunwindstack_callstack =
-      libunwindstack_result.frames();
-
-  if (libunwindstack_callstack.empty()) {
-    ORBIT_ERROR(
-        "Discarding sample as DWARF-based unwinding resulted in empty callstack (error: %s)",
-        LibunwindstackUnwinder::LibunwindstackErrorString(libunwindstack_result.error_code()));
-    return Callstack::kStackTopDwarfUnwindingError;
-  }
+      std::min<uint64_t>(stack_size, stack_dump_size_), true, /*max_frames=*/1);
+
+  // `PatchCallerOfLeafFunction` must not be called for a sample in the outermost frame (_start or
+  // __libc_start_main), so unwinding must not success with max_frames=1.
+  ORBIT_CHECK(!libunwindstack_result.IsSuccess());
```

Comment:
I'm a bit confused about this check. Where is it guaranteed that we are not calling `PatchCallerOfLeafFunction` for a sample in the outermost frame? And are we sure that libunwindstack couldn't "accidentally" return success in some special case?